[GeoMechanicsApplication] Line interface geometry can now generate edges

avdg81 opened the pull request:
**📝 Description**

The new line interface geometry now supports generating its edges. Note the that first edge's points are identical to the points of the first side. The second edge's points reference the points at the second side. However, since this edge has opposite orientation, the end points will be swapped and any intermediate points will be reversed, too.

avdg81 committed:
Any line interface geometry has two edges

avdg81 committed:
Edges of line interface element have opposite orientations

The nodes of the first edge are identical to the nodes of the first side. The nodes of the second edge have the end nodes swapped and any high-order nodes are reversed.

avdg81 committed:
Explained about the edges of the line interface geometry

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_geometries/test_interface_geometry.cpp:
```diff
@@ -363,6 +374,22 @@ KRATOS_TEST_CASE_IN_SUITE(GetLocalCoordinatesOfAllNodesOfThreePlusThreeNodedLine
+    KRATOS_EXPECT_EQ(edges_of_3_plus_3_geometry.size(), 2);
+    AssertNodeIdsOfGeometry(edges_of_3_plus_3_geometry(0), {1, 2, 3});
+    AssertNodeIdsOfGeometry(edges_of_3_plus_3_geometry(1), {5, 4, 6});
+}
```
Following the other tests (where 2+2 and 3+3 cases are tested separately), and considering there is no common setup for the two cases, I'd split this into two separate tests.

avdg81 replied:
The reason I've put them in a single test case is that the expected behavior is the same for all line interface geometries. But you rightly pointed out that we have the same split in all other test cases, too. So I have adopted your suggestion.

rfaasse commented on applications/GeoMechanicsApplication/custom_geometries/line_interface_geometry.h:
```diff
@@ -208,6 +208,26 @@ class LineInterfaceGeometry : public Geometry<Node>
+        const auto points = this->Points();
+
+        // The first edge coincides with the first side of the element
+        auto begin_of_second_side = points.ptr_begin() + (points.size() / 2);
```
I'm not sure due to the use, but could this be const auto?

avdg81 replied:
As far as I know, iterators are generally not made `const`, since the distinction is in the type of iterator used (`iterator` vs. `const_iterator`). In that sense, it is similar to pointers: we do declare pointers to mutable and immutable objects (i.e. `T*` and `const T*`), but rarely do we make the pointer itself `const` (i.e. `T* const` vs `const T* const`). So I would prefer to keep this `auto` rather than changing this to `const auto`. Is that acceptable to you?

rfaasse replied:
Ah yes good point, thanks for clarifying, let's leave it as is!

rfaasse commented on applications/GeoMechanicsApplication/custom_geometries/README.md:
```diff
@@ -15,4 +15,6 @@ Similarly, the 3+3 line interface geometry has the following node numbering for
\ No newline at end of file
+One thing to note, is that this line interface geometry does not implement functions from the Geometry base class which are related to the integration scheme. That is because most of the time, interface geometries are used with a Lobatto integration scheme, which is not supported by the Geometry base class.
+
+Any line interface geometry has two edges. The first edge coincides with the first (i.e. bottom) side of the geometry. The first edge's nodes are identical to the nodes at the first edge. For the 3+3 line interface geometry this means the list of node IDs equals [1, 2, 3]. The second edge references the nodes of the second (i.e. top) side of the line interface geometry. However, this edge travers ...
```
For the i.e. bottom/top, I would refer to the schematics specifically, since of course it's very arbitrary in a real model or a closed interface what's bottom/top.

avdg81 replied:
Yes, I implicitly referred to the bottom and top sides of the interface geometries as illustrated by the figures. I will make that more explicit.

rfaasse reviewed (commented):
Clear implementation that should also work for higher order mid geometries! Also nicely documented and tested, I only have a few minor suggestions.

avdg81 committed:
Processed Richard's review comments

- Split a unit test case to be consistent with other test cases.
- Be more explicit about the top and bottom sides of the line interface geometry in the documentation.

rfaasse reviewed (dismissed):
Thanks for incorporating the changes, looks good to go to me!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_geometries/line_interface_geometry.h:
```diff
@@ -208,6 +208,26 @@ class LineInterfaceGeometry : public Geometry<Node>
+        const auto nodes_of_first_edge = PointerVector<Node>{points.ptr_begin(), begin_of_second_side};
+
+        // The second edge coincides with the second side of the element. However, the nodes must be
+        // traversed in opposite direction.
```
" However, the nodes must be traversed in opposite direction." 
true, but maybe not directly clear.
"However, the node sequence is in reversed order." Is clearer to me, I wonder what others think.

avdg81 replied:
I have opted for this wording, since it's not simply a matter of reversing the nodes at the second side. The difficulty is in the fact that we first list the end nodes, and only then any high-order nodes. Simply reversing the list of nodes is incorrect. Consider the case where the end nodes have IDs 4 and 5, and the mid-side node has ID 6. Then from the element's point of view the node order is [4, 5, 6]. Reversing that would give [6, 5, 4], which is incorrect since the mid-side node is now at the front of the list. What I meant to say is if we would traverse this side from the edge's viewpoint, it is in opposite direction compared to what we do for the element itself. Perhaps we can discuss this further in person to see which wording would be clearest? Thanks.

WPK4FEM replied:
Can we then speak about opposite orientation ( with node number according to the opposite orientation )?

avdg81 replied:
Thank you for your suggestion. I have updated the documentation accordingly.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_geometries/line_interface_geometry.h:
```diff
@@ -208,6 +208,26 @@ class LineInterfaceGeometry : public Geometry<Node>
+        // The second edge coincides with the second side of the element. However, the nodes must be
+        // traversed in opposite direction.
+        auto nodes_of_second_edge = PointerVector<Node>{begin_of_second_side, points.ptr_end()};
+        std::swap(nodes_of_second_edge(0), nodes_of_second_edge(1)); // end nodes
+        std::reverse(nodes_of_second_edge.ptr_begin() + 2, nodes_of_second_edge.ptr_end()); // any high-order nodes
```
Nice, should indeed work generically. But in the tests have 0 or 1 midpoints so I doubt if this really is tested.

avdg81 replied:
You are right, the actual reversal of the high-order nodes is _not_ covered by the unit tests. In fact, the current tests would still pass if we were to take out the reversal of the high-order nodes. The reason I have added it, is to make the implementation complete and understandable. Otherwise, we might get surprising results when we start to make 4+4 or 5+5 line interface elements available. What are your thoughts on this?

WPK4FEM replied:
Leave as is for now, We won't make higher order interface elements without unit tests, so this will be handled in future.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_geometries/line_interface_geometry.h:
```diff
@@ -208,6 +208,26 @@ class LineInterfaceGeometry : public Geometry<Node>
+        std::reverse(nodes_of_second_edge.ptr_begin() + 2, nodes_of_second_edge.ptr_end()); // any high-order nodes
+
+        auto result = GeometriesArrayType{};
+        result.push_back(std::make_shared<MidGeometryType>(nodes_of_first_edge));
+        result.push_back(std::make_shared<MidGeometryType>(nodes_of_second_edge));
```
Why MidGeometryType, isn't just a LineGeometryType sufficient?

avdg81 replied:
No, the type of the mid-line geometry is defined by template parameter `MidGeometryType`. This generalization makes the extension towards plane interface elements quite straightforward.

WPK4FEM reviewed (commented):
Nice and small. 
I think it tests the current configurations, but needs more testing as soon as even higher order edges appear.

avdg81 committed:
Processed Wijtze Pieter's review comments

Clarified the opposite orientation of the geometry's second edge.

WPK4FEM reviewed (approved):
Good to go.
